tests/drivers/flash/nrf_qspi_nor/src/main.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <drivers/flash.h>
+
+#if (CONFIG_NORDIC_QSPI_NOR - 0)
+#define FLASH_DEVICE DT_LABEL(DT_INST(0, nordic_qspi_nor))
+#define FLASH_NAME "JEDEC QSPI-NOR"
+#endif
+#define FLASH_TEST_REGION_OFFSET 0xff000
+#define FLASH_SECTOR_SIZE        4096
+
+#define LENGTH		17
+#define EXPECTED_SIZE	256
+
+struct device *flash_dev;
+int rc;
+static uint8_t expected[EXPECTED_SIZE];
+
+static void test_setup(void)
+{
+	flash_dev = device_get_binding(FLASH_DEVICE);
+	for (int i = 0; i < EXPECTED_SIZE; i++) {
+		expected[i] = i;
+	}
+}
+
+static void test_erase(void)
+{
+	zassert_not_null(flash_dev, "SPI flash driver " FLASH_DEVICE
+			 " was not found!\n");
+
+	rc = flash_erase(flash_dev, FLASH_TEST_REGION_OFFSET,
+			 FLASH_SECTOR_SIZE);
+	zassert_equal(rc, 0, "Flash memory not properly erased");
+}
+
+static void test_read_after_erase(void)
+{
+	uint8_t buf[4] = {0xff, 0xff, 0xff, 0xff};
+	uint8_t read_buf[4];
+
+	rc = flash_read(flash_dev, FLASH_TEST_REGION_OFFSET,
+			read_buf, 4);
+	zassert_equal(rc, 0, "Cannot read flash");
+	zassert_equal(memcmp(buf, read_buf, 4), 0,
+		      "Flash memory not empty");
+}
+
+static void test_read_unaligned_address(void)
+{
+	const size_t len = sizeof(expected);
+	uint8_t buf[len];
+
+	rc = flash_write(flash_dev,
+			 FLASH_TEST_REGION_OFFSET,
+			 expected, EXPECTED_SIZE);
+	zassert_equal(rc, 0, "Cannot write to flash");
+
+	/* address offset */
+	for (off_t ad_o = 0; ad_o < 4; ad_o++) {
+		/* buffer offset */
+		for (off_t buf_o = 0; buf_o < 4; buf_o++) {
+			rc = flash_read(flash_dev,
```

Comment:
The buffer should be cleared before the read. To avoid taking the previous read result as the current one if for some reason the current read request is for example ignored.

Reply:
Added.